File: code/packages/rust/intel4004-gatelevel/src/registers.rs
//! Register file -- 16 x 4-bit registers built from D flip-flops.
//!
//! # How registers work in hardware
//!
//! A register is a group of D flip-flops that share a clock signal. Each
//! flip-flop stores one bit. A 4-bit register has 4 flip-flops. The Intel
//! 4004 has 16 such registers (R0-R15), for a total of 64 flip-flops just
//! for the register file.
//!
//! In this simulation, each register call goes through:
//!
//! ```text
//! data bits -> D flip-flop x 4 -> output bits
//! ```
//!
//! The flip-flops are edge-triggered: they capture new data on the rising
//! edge of the clock. Between edges, the stored value is stable.
//!
//! # Register pairs
//!
//! The 4004 organizes its 16 registers into 8 pairs:
//!
//! ```text
//! P0 = R0:R1, P1 = R2:R3, ..., P7 = R14:R15
//! ```
//!
//! A register pair holds an 8-bit value (high nibble in even register,
//! low nibble in odd register). Pairs are used for FIM, SRC, FIN, JIN.
//!
//! # Accumulator
//!
//! The accumulator is a separate 4-bit register, not part of the R0-R15
//! file. It has its own dedicated flip-flops and is connected directly to
//! the ALU's output bus.

use logic_gates::sequential::{register, FlipFlopState};

use crate::bits::{bits_to_int, int_to_bits};

/// 16 x 4-bit register file built from D flip-flops.
///
/// Each of the 16 registers is a group of 4 D flip-flops from the
/// logic_gates sequential module. Reading and writing go through
/// actual flip-flop state transitions.
pub struct RegisterFile {
    /// 16 registers, each with 4 flip-flop states (one per bit).
    states: Vec<Vec<FlipFlopState>>,
}

impl RegisterFile {
    /// Initialize 16 registers, each with 4-bit flip-flop state set to 0.
    pub fn new() -> Self {
        let mut states = Vec::with_capacity(16);
        for _ in 0..16 {
            // Initialize state by clocking zeros through
            let mut state: Vec<FlipFlopState> =
                (0..4).map(|_| FlipFlopState::default()).collect();
            register(&[0, 0, 0, 0], 0, &mut state);
            register(&[0, 0, 0, 0], 1, &mut state);
            states.push(state);
        }
        Self { states }
    }

    /// Read a register value. Returns 4-bit integer (0-15).
    ///
    /// In real hardware, this would route through a 16-to-1 multiplexer
    /// built from gates. We simulate the flip-flop read directly.
    pub fn read(&self, index: usize) -> u8 {
        // Read current output from flip-flops (clock=0, no write).
        // We clone the state because register() takes &mut but we don't
        // want to modify state for a read.
        let mut state = self.states[index].clone();
        let output = register(&[0, 0, 0, 0], 0, &mut state);
        bits_to_int(&output) as u8
    }

    /// Write a 4-bit value to a register.
    ///
    /// In real hardware: decoder selects the register, data bus presents
    /// the value, clock edge latches it into the flip-flops.
    pub fn write(&mut self, index: usize, value: u8) {
        let bits = int_to_bits((value & 0xF) as u16, 4);
        // Clock low (setup)
        register(&bits, 0, &mut self.states[index]);
        // Clock high (capture on rising edge)
        register(&bits, 1, &mut self.states[index]);
    }

    /// Read an 8-bit value from a register pair.
    ///
    /// Pair 0 = R0:R1 (R0=high nibble, R1=low nibble).
    pub fn read_pair(&self, pair_index: usize) -> u8 {
        let high = self.read(pair_index * 2);
        let low = self.read(pair_index * 2 + 1);
        (high << 4) | low
    }

    /// Write an 8-bit value to a register pair.
    pub fn write_pair(&mut self, pair_index: usize, value: u8) {
        self.write(pair_index * 2, (value >> 4) & 0xF);
        self.write(pair_index * 2 + 1, value & 0xF);
    }

    /// Reset all registers to 0 by clocking in zeros.
    pub fn reset(&mut self) {
        for i in 0..16 {
            self.write(i, 0);
        }
    }

    /// Gate count for the register file.
    ///
    /// 16 registers x 4 bits x ~6 gates per D flip-flop = 384 gates.
    /// Plus 4-to-16 decoder for write select: ~32 gates.
    /// Plus 16-to-1 mux for read select: ~64 gates.
    /// Total: ~480 gates.
    pub fn gate_count(&self) -> usize {
        480
    }
}

impl Default for RegisterFile {
    fn default() -> Self {
        Self::new()
    }
}

/// 4-bit accumulator register built from D flip-flops.
///
/// The accumulator is the 4004's main working register. Almost every
/// arithmetic and logic operation reads from or writes to it.
pub struct Accumulator {
    state: Vec<FlipFlopState>,
}

impl Accumulator {
    /// Initialize accumulator to 0.
    pub fn new() -> Self {
        let mut state: Vec<FlipFlopState> =
            (0..4).map(|_| FlipFlopState::default()).collect();
        register(&[0, 0, 0, 0], 0, &mut state);
        register(&[0, 0, 0, 0], 1, &mut state);
        Self { state }
    }

    /// Read the accumulator value (0-15).
    pub fn read(&self) -> u8 {
        let mut state = self.state.clone();
        let output = register(&[0, 0, 0, 0], 0, &mut state);
        bits_to_int(&output) as u8
    }

    /// Write a 4-bit value to the accumulator.
    pub fn write(&mut self, value: u8) {
        let bits = int_to_bits((value & 0xF) as u16, 4);
        register(&bits, 0, &mut self.state);
        register(&bits, 1, &mut self.state);
    }

    /// Reset to 0.
    pub fn reset(&mut self) {
        self.write(0);
    }

    /// 4 D flip-flops x ~6 gates = 24 gates.
    pub fn gate_count(&self) -> usize {
        24
    }
}

impl Default for Accumulator {
    fn default() -> Self {
        Self::new()
    }
}

/// 1-bit carry/borrow flag built from a D flip-flop.
///
/// The carry flag is set by arithmetic operations and read by
/// conditional jumps and multi-digit BCD arithmetic.
pub struct CarryFlag {
    state: Vec<FlipFlopState>,
}

impl CarryFlag {
    /// Initialize carry to false (0).
    pub fn new() -> Self {
        let mut state: Vec<FlipFlopState> =
            (0..1).map(|_| FlipFlopState::default()).collect();
        register(&[0], 0, &mut state);
        register(&[0], 1, &mut state);
        Self { state }
    }

    /// Read carry flag as a boolean.
    pub fn read(&self) -> bool {
        let mut state = self.state.clone();
        let output = register(&[0], 0, &mut state);
        output[0] == 1
    }

    /// Write carry flag.
    pub fn write(&mut self, value: bool) {
        let bit = if value { 1u8 } else { 0u8 };
        register(&[bit], 0, &mut self.state);
        register(&[bit], 1, &mut self.state);
    }

    /// Reset to false.
    pub fn reset(&mut self) {
        self.write(false);
    }

    /// 1 D flip-flop x ~6 gates = 6 gates.
    pub fn gate_count(&self) -> usize {
        6
    }
}

impl Default for CarryFlag {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_register_file_write_read() {
        let mut regs = RegisterFile::new();
        regs.write(3, 0xA);
        assert_eq!(regs.read(3), 0xA);
    }

    #[test]
    fn test_register_file_all_zeros() {
        let regs = RegisterFile::new();
        for i in 0..16 {
            assert_eq!(regs.read(i), 0);
        }
    }

    #[test]
    fn test_register_file_pair() {
        let mut regs = RegisterFile::new();
        regs.write_pair(2, 0xAB); // R4=0xA, R5=0xB
        assert_eq!(regs.read(4), 0xA);
        assert_eq!(regs.read(5), 0xB);
        assert_eq!(regs.read_pair(2), 0xAB);
    }

    #[test]
    fn test_register_file_reset() {
        let mut regs = RegisterFile::new();
        regs.write(0, 5);
        regs.write(15, 10);
        regs.reset();
        assert_eq!(regs.read(0), 0);
        assert_eq!(regs.read(15), 0);
    }

    #[test]
    fn test_accumulator() {
        let mut acc = Accumulator::new();
        assert_eq!(acc.read(), 0);
        acc.write(7);
        assert_eq!(acc.read(), 7);
        acc.write(15);
        assert_eq!(acc.read(), 15);
        acc.reset();
        assert_eq!(acc.read(), 0);
    }

    #[test]
    fn test_carry_flag() {
        let mut carry = CarryFlag::new();
        assert!(!carry.read());
        carry.write(true);
        assert!(carry.read());
        carry.write(false);
        assert!(!carry.read());
        carry.reset();
        assert!(!carry.read());
    }
}

File: code/packages/go/intel8008-gatelevel/registers.go
package intel8008gatelevel

// Register file — 7 × 8-bit registers built from D flip-flops.
//
// # The 8008's register file
//
// The Intel 8008 has 7 general-purpose 8-bit registers: A (accumulator),
// B, C, D, E, H, and L. Together these form 7 × 8 = 56 D flip-flops,
// compared to the 4004's 16 × 4 = 64 flip-flops (the 8008's register file
// actually has fewer flip-flops despite being an "upgrade" — the 4004 wasted
// transistors on register encoding flexibility).
//
// Each register write traverses:
//
//	value → IntToBits(value, 8) → Register(bits, clockEdge, state, width=8)
//	  → 8 × D_flip_flop(bit, clockEdge)
//	    → 8 × SR_latch(D, NOT(D))
//	      → 8 × 2 NOR gates = 16 NOR gates per register
//
// Total: 7 registers × 16 NOR gates = 112 NOR gates = 448 transistors
// (compare to the 4004's 16 × 8 = 128 NOR gates = 512 transistors)
//
// # Register encoding (3-bit SSS/DDD field)
//
// The 8008 uses a 3-bit field to identify registers in opcodes:
//
//	0 = B    1 = C    2 = D    3 = E
//	4 = H    5 = L    6 = M (pseudo-register, memory access)    7 = A
//
// Register index 6 is NOT a real register — it's the M pseudo-register
// that routes reads/writes to memory at the address (H & 0x3F) << 8 | L.
// Reading or writing reg 6 via this file is a programming error.
//
// # Flag register
//
// The 4 CPU flags (Carry, Zero, Sign, Parity) are stored in a 4-bit register
// of D flip-flops, separate from the data register file.

import (
	logicgates "github.com/adhithyan15/coding-adventures/code/packages/go/logic-gates"
)

// RegisterFile is a 7 × 8-bit register file built from D flip-flops.
//
// Indices 0-5 are B, C, D, E, H, L. Index 7 is A (accumulator).
// Index 6 is the M pseudo-register and is NOT stored here.
type RegisterFile struct {
	// 8 slots allocated (indices 0-7), index 6 is unused (M pseudo-register)
	// Each slot is a list of 8 FlipFlopState values (one per bit)
	states [8][]logicgates.FlipFlopState
}

// NewRegisterFile creates a new register file with all registers set to 0.
//
// Each register is initialized by clocking a zero byte through 8 flip-flops.
func NewRegisterFile() *RegisterFile {
	result, _ := StartNew[*RegisterFile]("intel8008-gatelevel.NewRegisterFile", nil,
		func(op *Operation[*RegisterFile], rf *ResultFactory[*RegisterFile]) *OperationResult[*RegisterFile] {
			regFile := &RegisterFile{}
			zeros := make([]int, 8)
			for i := 0; i < 8; i++ {
				if i == 6 {
					continue // Index 6 is M pseudo-register, skip
				}
				// Initialize: clock low (setup) then clock high (latch)
				_, state := logicgates.Register(zeros, 0, nil)
				_, state = logicgates.Register(zeros, 1, state)
				regFile.states[i] = state
			}
			return rf.Generate(true, false, regFile)
		}).GetResult()
	return result
}

// Read reads an 8-bit register value.
//
// index must be 0-7 (excluding 6 = M pseudo-register).
// In real hardware, this routes through a 3-to-8 decoder and an 8-to-1 mux.
// We simulate the flip-flop read directly.
func (regFile *RegisterFile) Read(index int) int {
	result, _ := StartNew[int]("intel8008-gatelevel.RegisterFile.Read", 0,
		func(op *Operation[int], rf *ResultFactory[int]) *OperationResult[int] {
			op.AddProperty("index", index)
			if index == 6 {
				panic("intel8008-gatelevel: register index 6 is M pseudo-register — use memory, not RegisterFile.Read")
			}
			// Read current flip-flop outputs (clock=0 = read mode, no write)
			zeros := make([]int, 8)
			output, _ := logicgates.Register(zeros, 0, regFile.states[index])
			return rf.Generate(true, false, BitsToInt(output))
		}).PanicOnUnexpected().GetResult()
	return result
}

// Write writes an 8-bit value to a register.
//
// In real hardware: the 3-bit register address is decoded by a 3-to-8
// decoder (8 AND/NOT gates), which enables the write path for exactly one
// register. The 8-bit data then clocks through 8 D flip-flops.
func (regFile *RegisterFile) Write(index int, value int) {
	_, _ = StartNew[struct{}]("intel8008-gatelevel.RegisterFile.Write", struct{}{},
		func(op *Operation[struct{}], rf *ResultFactory[struct{}]) *OperationResult[struct{}] {
			op.AddProperty("index", index)
			op.AddProperty("value", value)
			if index == 6 {
				panic("intel8008-gatelevel: register index 6 is M pseudo-register — use memory, not RegisterFile.Write")
			}
			bits := IntToBits(value&0xFF, 8)
			// Two-phase clocking: setup (clock=0) then capture (clock=1)
			_, state := logicgates.Register(bits, 0, regFile.states[index])
			_, state = logicgates.Register(bits, 1, state)
			regFile.states[index] = state
			return rf.Generate(true, false, struct{}{})
		}).PanicOnUnexpected().GetResult()
}

// HLAddress computes the 14-bit M pseudo-register address from H and L.
//
// The 8008 forms the memory address as:
//
//	address = (H & 0x3F) << 8 | L
//
// Why mask H to 6 bits? The 8008's address bus is 14 bits wide. H contributes
// the upper 6 bits (bits 8-13) and L contributes the lower 8 bits (bits 0-7).
// The upper 2 bits of H (bits 6-7) are ignored.
//
// In gate logic, this is a combination of AND gates (to mask H's upper bits),
// a shift (hardwired — no gates needed), and OR gates to combine.
func (regFile *RegisterFile) HLAddress() int {
	result, _ := StartNew[int]("intel8008-gatelevel.RegisterFile.HLAddress", 0,
		func(op *Operation[int], rf *ResultFactory[int]) *OperationResult[int] {
			h := regFile.Read(4) // H = register index 4
			l := regFile.Read(5) // L = register index 5

			// Mask H to 6 bits using AND gates: H & 0x3F
			hBits := IntToBits(h, 8)
			maskBits := IntToBits(0x3F, 8)
			hMasked := 0
			for i := 0; i < 8; i++ {
				hMasked |= logicgates.AND(hBits[i], maskBits[i]) << i
			}

			// Combine: address = (hMasked << 8) | l
			address := (hMasked << 8) | l
			return rf.Generate(true, false, address&0x3FFF) // ensure 14-bit
		}).GetResult()
	return result
}

// Reset resets all registers to 0.
func (regFile *RegisterFile) Reset() {
	_, _ = StartNew[struct{}]("intel8008-gatelevel.RegisterFile.Reset", struct{}{},
		func(op *Operation[struct{}], rf *ResultFactory[struct{}]) *OperationResult[struct{}] {
			for i := 0; i < 8; i++ {
				if i == 6 {
					continue // Skip M pseudo-register
				}
				regFile.Write(i, 0)
			}
			return rf.Generate(true, false, struct{}{})
		}).GetResult()
}

// GateCount returns the gate count for the register file.
//
// 7 registers × 8 bits × ~4 NOR gates per flip-flop = 224 NOR gates.
// Plus 3-to-8 decoder for write select: ~24 gates.
// Plus 8-to-1 mux for read select: ~56 gates.
// Total: ~304 gates.
func (regFile *RegisterFile) GateCount() int {
	result, _ := StartNew[int]("intel8008-gatelevel.RegisterFile.GateCount", 0,
		func(op *Operation[int], rf *ResultFactory[int]) *OperationResult[int] {
			return rf.Generate(true, false, 304)
		}).GetResult()
	return result
}

// FlagRegister stores the 4 CPU flags in a 4-bit D flip-flop register.
//
// Flags:
//
//	bit 0 = Carry
//	bit 1 = Zero
//	bit 2 = Sign
//	bit 3 = Parity (even parity → 1)
//
// This is 4 D flip-flops = ~16 NOR gates = ~64 transistors.
type FlagRegister struct {
	state []logicgates.FlipFlopState
}

// NewFlagRegister creates a flag register initialized to all zeros.
func NewFlagRegister() *FlagRegister {
	result, _ := StartNew[*FlagRegister]("intel8008-gatelevel.NewFlagRegister", nil,
		func(op *Operation[*FlagRegister], rf *ResultFactory[*FlagRegister]) *OperationResult[*FlagRegister] {
			zeros := make([]int, 4)
			_, state := logicgates.Register(zeros, 0, nil)
			_, state = logicgates.Register(zeros, 1, state)
			return rf.Generate(true, false, &FlagRegister{state: state})
		}).GetResult()
	return result
}

// ReadFlags reads all 4 flags as a packed integer.
//
// Returns bit 0=Carry, bit 1=Zero, bit 2=Sign, bit 3=Parity.
func (f *FlagRegister) ReadFlags() (carry, zero, sign, parity bool) {
	type flagVals struct {
		carry, zero, sign, parity bool
	}
	fv, _ := StartNew[flagVals]("intel8008-gatelevel.FlagRegister.ReadFlags", flagVals{},
		func(op *Operation[flagVals], rf *ResultFactory[flagVals]) *OperationResult[flagVals] {
			zeros := make([]int, 4)
			output, _ := logicgates.Register(zeros, 0, f.state)
			return rf.Generate(true, false, flagVals{
				carry:  output[0] == 1,
				zero:   output[1] == 1,
				sign:   output[2] == 1,
				parity: output[3] == 1,
			})
		}).GetResult()
	return fv.carry, fv.zero, fv.sign, fv.parity
}

// WriteFlags writes all 4 flags.
func (f *FlagRegister) WriteFlags(carry, zero, sign, parity bool) {
	_, _ = StartNew[struct{}]("intel8008-gatelevel.FlagRegister.WriteFlags", struct{}{},
		func(op *Operation[struct{}], rf *ResultFactory[struct{}]) *OperationResult[struct{}] {
			bits := make([]int, 4)
			if carry {
				bits[0] = 1
			}
			if zero {
				bits[1] = 1
			}
			if sign {
				bits[2] = 1
			}
			if parity {
				bits[3] = 1
			}
			_, state := logicgates.Register(bits, 0, f.state)
			_, state = logicgates.Register(bits, 1, state)
			f.state = state
			return rf.Generate(true, false, struct{}{})
		}).GetResult()
}

// Reset resets all flags to false.
func (f *FlagRegister) Reset() {
	f.WriteFlags(false, false, false, false)
}

// GateCount returns ~16 gates (4 flip-flops × ~4 NOR gates each).
func (f *FlagRegister) GateCount() int {
	result, _ := StartNew[int]("intel8008-gatelevel.FlagRegister.GateCount", 0,
		func(op *Operation[int], rf *ResultFactory[int]) *OperationResult[int] {
			return rf.Generate(true, false, 16)
		}).GetResult()
	return result
}
